File: client/src/api/power-presets.ts
/**
 * API client for Power Presets
 *
 * Manages power presets that define CPU clock speeds for each service power property.
 */

import { apiClient } from '../lib/api';

// Types
export type ServicePowerProperty = 'idle' | 'low' | 'medium' | 'surge';

export interface PowerPreset {
  id: number;
  name: string;
  description?: string;
  is_system_preset: boolean;
  is_active: boolean;
  base_clock_mhz: number;
  idle_clock_mhz: number;
  low_clock_mhz: number;
  medium_clock_mhz: number;
  surge_clock_mhz: number;
  created_at: string;
  updated_at: string;
}

export interface PowerPresetSummary {
  id: number;
  name: string;
  is_system_preset: boolean;
  is_active: boolean;
}

export interface PowerPresetListResponse {
  presets: PowerPreset[];
  active_preset?: PowerPreset;
}

export interface CreatePresetRequest {
  name: string;
  description?: string;
  base_clock_mhz?: number;
  idle_clock_mhz?: number;
  low_clock_mhz?: number;
  medium_clock_mhz?: number;
  surge_clock_mhz?: number;
}

export interface UpdatePresetRequest {
  name?: string;
  description?: string;
  base_clock_mhz?: number;
  idle_clock_mhz?: number;
  low_clock_mhz?: number;
  medium_clock_mhz?: number;
  surge_clock_mhz?: number;
}

export interface ActivatePresetResponse {
  success: boolean;
  message: string;
  previous_preset?: PowerPresetSummary;
  new_preset: PowerPresetSummary;
}

// Preset display info
export const PRESET_INFO: Record<string, { name: string; color: string; icon: string; description: string }> = {
  'Energy Saver': {
    name: 'Energy Saver',
    color: 'emerald',
    icon: '🌱',
    description: 'Minimaler Stromverbrauch',
  },
  'Balanced': {
    name: 'Balanced',
    color: 'blue',
    icon: '⚖️',
    description: 'Ausgewogene Balance',
  },
  'Performance': {
    name: 'Performance',
    color: 'red',
    icon: '🚀',
    description: 'Maximale Leistung',
  },
};

// Property display info
export const PROPERTY_INFO: Record<ServicePowerProperty, { name: string; color: string; icon: string; description: string }> = {
  idle: {
    name: 'Idle',
    color: 'emerald',
    icon: '🌙',
    description: 'Leerlauf, Monitoring',
  },
  low: {
    name: 'Low',
    color: 'blue',
    icon: '🔋',
    description: 'CRUD, Konfiguration',
  },
  medium: {
    name: 'Medium',
    color: 'yellow',
    icon: '⚡',
    description: 'File-Ops, Sync, SMART',
  },
  surge: {
    name: 'Surge',
    color: 'red',
    icon: '🔥',
    description: 'Backup, RAID Rebuild',
  },
};

// API Functions

/**
 * Get all power presets
 */
export async function listPresets(): Promise<PowerPresetListResponse> {
  const response = await apiClient.get<PowerPresetListResponse>('/api/power/presets/');
  return response.data;
}

/**
 * Get the currently active preset
 */
export async function getActivePreset(): Promise<PowerPreset> {
  const response = await apiClient.get<PowerPreset>('/api/power/presets/active');
  return response.data;
}

/**
 * Get a specific preset by ID
 */
export async function getPreset(presetId: number): Promise<PowerPreset> {
  const response = await apiClient.get<PowerPreset>(`/api/power/presets/${presetId}`);
  return response.data;
}

/**
 * Activate a preset (admin only)
 */
export async function activatePreset(presetId: number): Promise<ActivatePresetResponse> {
  const response = await apiClient.post<ActivatePresetResponse>(
    `/api/power/presets/${presetId}/activate`
  );
  return response.data;
}

/**
 * Create a new custom preset (admin only)
 */
export async function createPreset(data: CreatePresetRequest): Promise<PowerPreset> {
  const response = await apiClient.post<PowerPreset>('/api/power/presets/', data);
  return response.data;
}

/**
 * Update an existing preset (admin only)
 */
export async function updatePreset(
  presetId: number,
  data: UpdatePresetRequest
): Promise<PowerPreset> {
  const response = await apiClient.put<PowerPreset>(`/api/power/presets/${presetId}`, data);
  return response.data;
}

/**
 * Delete a custom preset (admin only)
 */
export async function deletePreset(presetId: number): Promise<{ success: boolean; message: string }> {
  const response = await apiClient.delete<{ success: boolean; message: string }>(
    `/api/power/presets/${presetId}`
  );
  return response.data;
}

/**
 * Get clock speed for a property from a preset
 */
export function getClockForProperty(preset: PowerPreset, property: ServicePowerProperty): number {
  switch (property) {
    case 'idle':
      return preset.idle_clock_mhz;
    case 'low':
      return preset.low_clock_mhz;
    case 'medium':
      return preset.medium_clock_mhz;
    case 'surge':
      return preset.surge_clock_mhz;
    default:
      return preset.base_clock_mhz;
  }
}

/**
 * Format clock speed for display
 */
export function formatClockSpeed(mhz: number): string {
  if (mhz >= 1000) {
    return `${(mhz / 1000).toFixed(1)} GHz`;
  }
  return `${mhz} MHz`;
}

File: backend/app/api/routes/power_presets.py
"""
Power Presets API Routes.

Provides endpoints for managing power presets that define
CPU clock speeds for each service power property level.
"""

from typing import List

from fastapi import APIRouter, Depends, HTTPException, Request, Response, status

from app.api import deps
from app.core.rate_limiter import user_limiter, get_limit
from app.schemas.power import (
    PowerPresetCreate,
    PowerPresetUpdate,
    PowerPresetResponse,
    PowerPresetListResponse,
    PowerPresetSummary,
    ActivatePresetResponse,
)
from app.schemas.user import UserPublic
from app.services.power_preset_service import get_preset_service

router = APIRouter(prefix="/power/presets", tags=["power-presets"])


@router.get("/", response_model=PowerPresetListResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def list_presets(
    request: Request, response: Response,
    _: UserPublic = Depends(deps.get_current_user)
) -> PowerPresetListResponse:
    """
    List all power presets.

    Returns both system presets (Energy Saver, Balanced, Performance)
    and any custom presets created by admins.
    """
    service = get_preset_service()
    presets = await service.list_presets()
    active = await service.get_active_preset()

    preset_responses = [
        PowerPresetResponse(
            id=p.id,
            name=p.name,
            description=p.description,
            is_system_preset=p.is_system_preset,
            is_active=p.is_active,
            base_clock_mhz=p.base_clock_mhz,
            idle_clock_mhz=p.idle_clock_mhz,
            low_clock_mhz=p.low_clock_mhz,
            medium_clock_mhz=p.medium_clock_mhz,
            surge_clock_mhz=p.surge_clock_mhz,
            created_at=p.created_at,
            updated_at=p.updated_at,
        )
        for p in presets
    ]

    active_response = None
    if active:
        active_response = PowerPresetResponse(
            id=active.id,
            name=active.name,
            description=active.description,
            is_system_preset=active.is_system_preset,
            is_active=active.is_active,
            base_clock_mhz=active.base_clock_mhz,
            idle_clock_mhz=active.idle_clock_mhz,
            low_clock_mhz=active.low_clock_mhz,
            medium_clock_mhz=active.medium_clock_mhz,
            surge_clock_mhz=active.surge_clock_mhz,
            created_at=active.created_at,
            updated_at=active.updated_at,
        )

    return PowerPresetListResponse(
        presets=preset_responses,
        active_preset=active_response
    )


@router.get("/active", response_model=PowerPresetResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def get_active_preset(
    request: Request, response: Response,
    _: UserPublic = Depends(deps.get_current_user)
) -> PowerPresetResponse:
    """
    Get the currently active power preset.
    """
    service = get_preset_service()
    preset = await service.get_active_preset()

    if preset is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No active preset found"
        )

    return PowerPresetResponse(
        id=preset.id,
        name=preset.name,
        description=preset.description,
        is_system_preset=preset.is_system_preset,
        is_active=preset.is_active,
        base_clock_mhz=preset.base_clock_mhz,
        idle_clock_mhz=preset.idle_clock_mhz,
        low_clock_mhz=preset.low_clock_mhz,
        medium_clock_mhz=preset.medium_clock_mhz,
        surge_clock_mhz=preset.surge_clock_mhz,
        created_at=preset.created_at,
        updated_at=preset.updated_at,
    )


@router.get("/{preset_id}", response_model=PowerPresetResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def get_preset(
    request: Request, response: Response,
    preset_id: int,
    _: UserPublic = Depends(deps.get_current_user)
) -> PowerPresetResponse:
    """
    Get a specific preset by ID.
    """
    service = get_preset_service()
    preset = await service.get_preset_by_id(preset_id)

    if preset is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Preset not found: {preset_id}"
        )

    return PowerPresetResponse(
        id=preset.id,
        name=preset.name,
        description=preset.description,
        is_system_preset=preset.is_system_preset,
        is_active=preset.is_active,
        base_clock_mhz=preset.base_clock_mhz,
        idle_clock_mhz=preset.idle_clock_mhz,
        low_clock_mhz=preset.low_clock_mhz,
        medium_clock_mhz=preset.medium_clock_mhz,
        surge_clock_mhz=preset.surge_clock_mhz,
        created_at=preset.created_at,
        updated_at=preset.updated_at,
    )


@router.post("/{preset_id}/activate", response_model=ActivatePresetResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def activate_preset(
    request: Request, response: Response,
    preset_id: int,
    user: UserPublic = Depends(deps.get_current_admin)
) -> ActivatePresetResponse:
    """
    Activate a power preset (admin only).

    The selected preset will define the CPU clock speeds used for
    each service power property level.
    """
    service = get_preset_service()

    # Get current active preset
    current = await service.get_active_preset()
    previous_summary = None
    if current:
        previous_summary = PowerPresetSummary(
            id=current.id,
            name=current.name,
            is_system_preset=current.is_system_preset,
            is_active=current.is_active
        )

    # Activate the new preset
    success = await service.activate_preset(preset_id)

    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Preset not found: {preset_id}"
        )

    # Get the newly activated preset
    new_preset = await service.get_preset_by_id(preset_id)
    new_summary = PowerPresetSummary(
        id=new_preset.id,
        name=new_preset.name,
        is_system_preset=new_preset.is_system_preset,
        is_active=True
    )

    return ActivatePresetResponse(
        success=True,
        message=f"Activated preset: {new_preset.name}",
        previous_preset=previous_summary,
        new_preset=new_summary
    )


@router.post("/", response_model=PowerPresetResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def create_preset(
    request: Request, response: Response,
    data: PowerPresetCreate,
    user: UserPublic = Depends(deps.get_current_admin)
) -> PowerPresetResponse:
    """
    Create a new custom power preset (admin only).

    Custom presets can be modified and deleted. They are useful
    for creating specialized configurations for specific workloads.
    """
    service = get_preset_service()
    preset = await service.create_preset(data)

    if preset is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Could not create preset. Name '{data.name}' may already exist."
        )

    return PowerPresetResponse(
        id=preset.id,
        name=preset.name,
        description=preset.description,
        is_system_preset=preset.is_system_preset,
        is_active=preset.is_active,
        base_clock_mhz=preset.base_clock_mhz,
        idle_clock_mhz=preset.idle_clock_mhz,
        low_clock_mhz=preset.low_clock_mhz,
        medium_clock_mhz=preset.medium_clock_mhz,
        surge_clock_mhz=preset.surge_clock_mhz,
        created_at=preset.created_at,
        updated_at=preset.updated_at,
    )


@router.put("/{preset_id}", response_model=PowerPresetResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def update_preset(
    request: Request, response: Response,
    preset_id: int,
    data: PowerPresetUpdate,
    user: UserPublic = Depends(deps.get_current_admin)
) -> PowerPresetResponse:
    """
    Update an existing power preset (admin only).

    Note: System presets can only have clock values updated,
    not their name or description.
    """
    service = get_preset_service()
    preset = await service.update_preset(preset_id, data)

    if preset is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Could not update preset {preset_id}. It may not exist or name may be duplicate."
        )

    return PowerPresetResponse(
        id=preset.id,
        name=preset.name,
        description=preset.description,
        is_system_preset=preset.is_system_preset,
        is_active=preset.is_active,
        base_clock_mhz=preset.base_clock_mhz,
        idle_clock_mhz=preset.idle_clock_mhz,
        low_clock_mhz=preset.low_clock_mhz,
        medium_clock_mhz=preset.medium_clock_mhz,
        surge_clock_mhz=preset.surge_clock_mhz,
        created_at=preset.created_at,
        updated_at=preset.updated_at,
    )


@router.delete("/{preset_id}")
@user_limiter.limit(get_limit("admin_operations"))
async def delete_preset(
    request: Request, response: Response,
    preset_id: int,
    user: UserPublic = Depends(deps.get_current_admin)
) -> dict:
    """
    Delete a custom power preset (admin only).

    System presets cannot be deleted.
    Active presets cannot be deleted.
    """
    service = get_preset_service()

    # Check if it's a system preset
    preset = await service.get_preset_by_id(preset_id)
    if preset is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Preset not found: {preset_id}"
        )

    if preset.is_system_preset:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot delete system presets"
        )

    if preset.is_active:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot delete active preset. Activate another preset first."
        )

    success = await service.delete_preset(preset_id)

    if not success:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to delete preset"
        )

    return {"success": True, "message": f"Deleted preset: {preset.name}"}
